cpp/src/cider/common/hashtable/HashTableKeyHolder.h

```diff
@@ -89,4 +90,52 @@ inline void ALWAYS_INLINE keyHolderPersistKey(Key&&) {}
  */
 template <typename Key>
 inline void ALWAYS_INLINE keyHolderDiscardKey(Key&&) {}
+
+/**
+ * ArenaKeyHolder is a key holder for hash tables that serializes a StringRef
+ * key to an Arena.
+ */
+struct ArenaKeyHolder {
+  StringRef key;
+  Arena& pool;
+};
+
+}  // namespace cider::hashtable
+
+inline StringRef& ALWAYS_INLINE keyHolderGetKey(ArenaKeyHolder& holder) {
+  return holder.key;
+}
+
+inline void ALWAYS_INLINE keyHolderPersistKey(ArenaKeyHolder& holder) {
+  // Hash table shouldn't ask us to persist a zero key
+  assert(holder.key.size > 0);
+  holder.key.data = holder.pool.insert(holder.key.data, holder.key.size);
+}
+
+inline void ALWAYS_INLINE keyHolderDiscardKey(ArenaKeyHolder&) {}
+
+namespace cider::hashtable {
+
+/** SerializedKeyHolder is a key holder for a StringRef key that is already
+ * serialized to an Arena. The key must be the last allocation in this Arena,
+ * and is discarded by rolling back the allocation.
+ */
+struct SerializedKeyHolder {
+  StringRef key;
+  Arena& pool;

```

Comment:
using pointer?

Reply:
The lifecycle of pool here should be along with `SerializedKeyHolder`

Reply:
I think we should use the value of `SerializedKeyHolder` instead of the pointer, and `SerializedKeyHolder` is a struct type.